common/utils/include/claragenomics/utils/device_preallocated_allocator.cuh

```diff
@@ -0,0 +1,306 @@
+
+        if (block_to_get_memory_from_iter == std::end(free_blocks_))
+        {
+            return cudaErrorMemoryAllocation;
+        }
+
+        MemoryBlock new_memory_block{block_to_get_memory_from_iter->begin,
+                                     bytes_needed,
+                                     associated_stream};
+
+        // ** reduce the size of the block the memory is going to be taken from
+        if (block_to_get_memory_from_iter->size == bytes_needed)
+        {
+            // this memory block is completely used, remove it
+            free_blocks_.erase(block_to_get_memory_from_iter);
+        }
+        else
+        {
+            // there will still be some memory left, update free block size
+            block_to_get_memory_from_iter->begin += new_memory_block.size;
+            block_to_get_memory_from_iter->size -= new_memory_block.size;
+        }
+
+        // ** add new used memory block to the list of used blocks
+
+        // look for the block right after the block that is to be added
+        auto insert_used_block_before_iter = std::find_if(std::begin(used_blocks_),
+                                                          std::end(used_blocks_),
+                                                          [&new_memory_block](const MemoryBlock& memory_block) {
+                                                              return memory_block.begin > new_memory_block.begin;
+                                                          });
+
+        // insert new block in the array
+        used_blocks_.insert(insert_used_block_before_iter, new_memory_block);
+
+        // set pointer to new location
+        *ptr = static_cast<void*>(buffer_ptr_.get() + new_memory_block.begin);
+        return cudaSuccess;
+    }
+
+    /// @brief returns the block starting at pointer
+    /// This function blocks until all work on associated_stream is done
+    /// @param pointer pointer at the begining of the block to be freed
+    /// @return error status
+    cudaError_t free_block(void* pointer)
+    {
+        const size_t block_start = static_cast<char*>(pointer) - buffer_ptr_.get();
+        assert(block_start >= 0);
+        assert(block_start < buffer_size_);
+
+        // ** look for pointer's memory block
+        auto block_to_be_freed_iter = std::find_if(std::begin(used_blocks_),
+                                                   std::end(used_blocks_),
+                                                   [block_start](const MemoryBlock& memory_block) {
+                                                       return memory_block.begin == block_start;
+                                                   });
+        assert(block_to_be_freed_iter != std::end(used_blocks_));
+
+        // ** wait for all work on associated_stream to finish before freeing up this memory block
+        CGA_CU_CHECK_ERR(cudaStreamSynchronize(block_to_be_freed_iter->associated_stream));
```

Comment:
Shouldn't the error of `cudaStreamSynchronize` be returned by the function? Or at least use `CGA_CU_ABORT_ON_ERR`?

Reply:
Good point, I forgot this function will be called from destructor. Changing to `CGA_CU_ABORT_ON_ERR`